dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -101,41 +102,54 @@ void FramesDecoder::InitAvState() {
     ret >= 0,
     make_string("Could not fill the codec based on parameters: ", detail::av_error_string(ret)));
 
-  ret = avcodec_open2(av_state_->codec_ctx_, av_state_->codec_, nullptr);
-  DALI_ENFORCE(
-    ret == 0,
-    make_string("Could not initialize codec context: ", detail::av_error_string(ret)));
-
-  av_state_->frame_ = av_frame_alloc();
-  DALI_ENFORCE(av_state_->frame_, "Could not allocate the av frame");
-
   av_state_->packet_ = av_packet_alloc();
   DALI_ENFORCE(av_state_->packet_, "Could not allocate av packet");
+
+  if (init_codecs) {
+    ret = avcodec_open2(av_state_->codec_ctx_, av_state_->codec_, nullptr);
+    DALI_ENFORCE(
+      ret == 0,
+      make_string("Could not initialize codec context: ", detail::av_error_string(ret)));
+
+    av_state_->frame_ = av_frame_alloc();
+    DALI_ENFORCE(av_state_->frame_, "Could not allocate the av frame");
+  }
 }
 
 bool FramesDecoder::CheckCodecSupport() {
   return std::find(
     SupportedCodecs.begin(),
     SupportedCodecs.end(),
-    av_state_->codec_->id) != SupportedCodecs.end();
+    av_state_->codec_params_->codec_id) != SupportedCodecs.end();
 }
 
-void FramesDecoder::FindVideoStream() {
-  for (size_t i = 0; i < av_state_->ctx_->nb_streams; ++i) {
-    av_state_->codec_params_ = av_state_->ctx_->streams[i]->codecpar;
-    av_state_->codec_ = avcodec_find_decoder(av_state_->codec_params_->codec_id);
+void FramesDecoder::FindVideoStream(bool init_codecs) {
+  if (init_codecs) {
+    for (size_t i = 0; i < av_state_->ctx_->nb_streams; ++i) {
+      av_state_->codec_params_ = av_state_->ctx_->streams[i]->codecpar;
+      av_state_->codec_ = avcodec_find_decoder(av_state_->codec_params_->codec_id);
```

Comment:
I know that this new code path with `init_codecs` is needed here, because this call fails for `AV_CODEC_ID_MPEG4`. Do you know why it does?

Reply:
Our build configuration doesn't list anything with mpeg in the name in the `Enabled decoders:` list. I guess we would need to use  "--enable-gpl" or "--enable-nonfree" but we can't due to the license.
If we move to a solution where we can use the libaviutils the user may have installed separately we can lose that limitation.
Still, this code path should be better as we avoid unnecessary codec initialization on the CPU.

Reply:
I will investigate, if we need "--enable-gpl" or "--enable-nonfree". We need it for `libxvid` but I am not 100% that we need it for native `mpeg4` decoder. Is that fine?

Reply:
I suggest to merge this PR and enable mpeg4 support for CPU separately as the GPU support is more important for us.

Reply:
https://github.com/NVIDIA/DALI/pull/4361 should cover both CPU and GPU MPEG-4 Part2 decoding.